dev/com.ibm.ws.config/src/com/ibm/ws/config/xml/internal/ConfigRefresher.java

```diff
@@ -136,7 +159,17 @@ private synchronized void doRefresh(Map<String, DeltaType> variableDelta) {
             // Let the notification show that we got an error while making the configuration changes
             configUpdatesDelivered.setResult(e);
         } finally {
-            changesEnded(configUpdatesDelivered);
+            if (CheckpointPhase.getPhase().restored() && changesEndedHook.runMultiThreaded) {
+                synchronized (changesEndedHook) {
+                    if (changesEndedHook.runMultiThreaded) {
+                        changesEndedHook.configUpdatesToDeliver.add(configUpdatesDelivered);
+                    } else {
+                        changesEnded(configUpdatesDelivered);
+                    }
+                }
+            } else {
+                changesEnded(configUpdatesDelivered);
+            }
```

Comment:
I think we should create a method `boolean ChangesEndedHook.queueNotification(RuntimeUpdateNotification)` method to call here instead of leaking `runMultiThreaded` and `configUpdatesToDeliver` out to here.  So it would end up being:

```java
            if (!changesEndedHook.queueNotification(configUpdatesDelivered) {
                changesEnded(configUpdatesDelivered);
            }
```

Then the queueNotifications could look like this:

```java
    private class ChangesEndedHook implements CheckpointHook {

        private final ThreadLocal<Boolean> checkpointThread = new ThreadLocal<Boolean>() {
             @Override protected Boolean initialValue() {
                 return Boolean.FALSE;
             }
         };

        // FIFO queue of deferred config update notifications
        final Deque<RuntimeUpdateNotification> configUpdatesToDeliver = new ArrayDeque<RuntimeUpdateNotification>();

        boolean queueNotification(RuntimeUpdateNotification notification) {
            if (!checkpointThreaded.get()) {
                return false;
            }
            configUpdatesToDeliver.add(notification);
        }

        @Override
        public void prepare() {
            checkpointThread.set(true);
        }

        @Override
        public void restore() {
            checkpointThread.set(false);
            RuntimeUpdateNotification notification = null;
            while (notification = configUpdatesToDeliver.pollFirst() != null) {
                changesEnded(notification);
            }
        }
    }
```